stan/math/opencl/matrix_cl.hpp

```diff
@@ -332,28 +332,47 @@ class matrix_cl<T, require_arithmetic_t<T>> {
 
   /**
    * Constructor for the matrix_cl that
-   * creates a copy of a scalar on the OpenCL device.
+   * creates a copy of a rvalue reference
+   * of a scalar on the OpenCL device.
    * Regardless of `partial_view`, whole matrix is stored.
    *
    * @param A the scalar
    * @param partial_view which part of the matrix is used
    */
-  explicit matrix_cl(const T& A,
+  template <typename Scal,
+            typename = require_same_t<T, std::remove_reference_t<Scal>>>
+  explicit matrix_cl(Scal&& A,
                      matrix_cl_view partial_view = matrix_cl_view::Diagonal)
       : rows_(1), cols_(1), view_(partial_view) {
     cl::Context& ctx = opencl_context.context();
     cl::CommandQueue& queue = opencl_context.queue();
     try {
-      buffer_cl_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T));
+      buffer_cl_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(std::decay_t<T>));
       cl::Event transfer_event;
-      queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T), &A, nullptr,
-                               &transfer_event);
+      queue.enqueueWriteBuffer(
+          buffer_cl_,
+          opencl_context.in_order() || std::is_rvalue_reference<T&&>::value, 0,
```

Comment:
Do we need the T&& here? If T is an rvalue it should have && already

Reply:
Yes. T is either value or lvalue reference, but never rvalue reference.

Reply:
We either use this or `std::is_rvalue_reference<decltype(A)>`, which I had before but then Tadej suggest T&& which I agree is nicer.